pkg/network/kafka/ebpf_main.go

```diff
@@ -0,0 +1,216 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+// +build linux_bpf
+
+package kafka
+
+import (
+	"fmt"
+	"math"
+	"os"
+
+	"github.com/cilium/ebpf"
+	"github.com/iovisor/gobpf/pkg/cpupossible"
+	"golang.org/x/sys/unix"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	manager "github.com/DataDog/ebpf-manager"
+)
+
+const (
+	kafkaInFlightMap                = "kafka_in_flight"
+	kafkaBatchesMap                 = "kafka_batches"
+	kafkaBatchStateMap              = "kafka_batch_state"
+	kafkaBatchEvents                = "kafka_batch_events"
+	kafkaLastTCPSeqPerConnectionMap = "kafka_last_tcp_seq_per_connection"
+
+	// ELF section of the BPF_PROG_TYPE_SOCKET_FILTER program used
+	// to inspect plain kafka traffic
+	kafkaSocketFilterStub = "socket/kafka_filter_entry"
+	kafkaSocketFilter     = "socket/kafka_filter"
+	kafkaProgsMap         = "kafka_progs"
+
+	// size of the channel containing completed kafka_notification_objects
+	batchNotificationsChanSize = 100
+
+	probeUID = "kafka"
+)
+
+type ebpfProgram struct {
+	*manager.Manager
+	cfg        *config.Config
+	bytecode   bytecode.AssetReader
+	offsets    []manager.ConstantEditor
+	mapCleaner *ddebpf.MapCleaner
+
+	batchCompletionHandler *ddebpf.PerfHandler
+}
+
+var tailCalls = []manager.TailCallRoute{
+	{
+		ProgArrayName: kafkaProgsMap,
+		Key:           kafkaProg,
+		ProbeIdentificationPair: manager.ProbeIdentificationPair{
+			EBPFSection:  kafkaSocketFilter,
+			EBPFFuncName: "socket__kafka_filter",
+		},
+	},
+}
+
+func newEBPFProgram(c *config.Config, offsets []manager.ConstantEditor) (*ebpfProgram, error) {
+	bc, err := getBytecode(c)
+	if err != nil {
+		return nil, err
+	}
+
+	batchCompletionHandler := ddebpf.NewPerfHandler(batchNotificationsChanSize)
+	mgr := &manager.Manager{
+		Maps: []*manager.Map{
+			{Name: kafkaInFlightMap},
+			{Name: kafkaLastTCPSeqPerConnectionMap},
+			{Name: kafkaBatchesMap},
+			{Name: kafkaBatchStateMap},
+		},
+		PerfMaps: []*manager.PerfMap{
+			{
+				Map: manager.Map{Name: kafkaBatchEvents},
+				PerfMapOptions: manager.PerfMapOptions{
+					PerfRingBufferSize: 256 * os.Getpagesize(),
```

Comment:
@p-lambert @brycekahle any reason to use `256` factor? I think @brycekahle asked to use a lower factor (`8`)

Reply:
Copied it from the HTTP data path, @p-lambert I would love to hear what's your opinion about it

Reply:
sorry for the delayed response. yes, that feels very excessive since this will essentially allocate 1Mb of (shared) memory per CPU core :grimacing:  (we should also lower it for HTTP). I think using a factor of 8 sounds good as 8 pages can fit ~85 `kafka_transaction_t` instances, which is probably enough to support an eventual burst of events

Reply:
did you consider https://github.com/DataDog/datadog-agent/pull/14707 in your calculations?

Reply:
nope, so it should be way more than 85 entries per CPU :sweat_smile:

Reply:
I might be missing anything
but the standard pagesize is 4k, the current struct size is 912 bytes, thus we can have (currently) 35 `kafka_transaction_t`
the change I'm introducing spares 320 bytes, which gives us the option to have 55 `kafka_transaction_t` instances in 8 pages